tensorly/decomposition/_tucker.py

```diff
@@ -294,7 +349,156 @@ def non_negative_tucker(tensor, rank, n_iter_max=10, init='svd', tol=10e-5,
+        Has to be set not to update a factor, 0 and 1 for U and V respectively
+        Default: None
+    verbose : boolean
+        Indicates whether the algorithm prints the successive
+        reconstruction errors or not
+        Default: False
+    return_errors : boolean
+        Indicates whether the algorithm should return all reconstruction errors
+        and computation time of each iteration or not
+        Default: False
+    exact : If it is True, the HALS nnls subroutines give results with high precision but it needs high computational cost. 
+        If it is False, the algorithm gives an approximate solution.
+        Default: False
+    algorithm : {'fista', 'as'}
+         Non negative least square solution to update the core. 
+         Default: 'fista'
+    Returns
+    -------
+    factors : ndarray list
+            list of positive factors of the CP decomposition
+            element `i` is of shape ``(tensor.shape[i], rank)``
+    errors: list
+        A list of reconstruction errors at each iteration of the algorithm.
+
+    References
+    ----------
+    .. [1] tl.G.Kolda and B.W.Bader, "Tensor Decompositions and Applications",
+       SIAM REVIEW, vol. 51, n. 3, pp. 455-500, 2009.
+    """
+    rank = validate_tucker_rank(tl.shape(tensor), rank=rank)
+    n_modes = tl.ndim(tensor)
+    if sparsity_coefficients is None or len(sparsity_coefficients) != n_modes:
+        sparsity_coefficients = [sparsity_coefficients] * n_modes
+
+    if fixed_modes is None:
+        fixed_modes = []
+
+    # Avoiding errors
+    for fixed_value in fixed_modes:
+        sparsity_coefficients[fixed_value] = None
+    # Generating the mode update sequence
+    modes = [mode for mode in range(tl.ndim(tensor)) if mode not in fixed_modes]
+
+    nn_core, nn_factors = initialize_tucker(tensor, rank, modes, init=init, svd=svd, random_state=random_state, 
+                                            non_negative=True)
+    # initialisation - declare local varaibles
+    norm_tensor = tl.norm(tensor, 2)
+    rec_errors = []
+
+    # Iterate over one step of NTF
+    for iteration in range(n_iter_max):
+        # One pass of least squares on each updated mode
+        for mode in modes:
+
+            # Computing Hadamard of cross-products
+            pseudo_inverse = nn_factors.copy()
+            for i, factor in enumerate(nn_factors):
+                if i != mode:
+                    pseudo_inverse[i] = tl.dot(tl.conj(tl.transpose(factor)), factor)
+            # Second, the multiway product with core G
+            core_cross = multi_mode_dot(nn_core, pseudo_inverse, skip=mode)
+            UtU = tl.dot(unfold(core_cross, mode), tl.transpose(unfold(nn_core, mode)))
+
+            # UtM
+            tensor_cross = multi_mode_dot(tensor, nn_factors, skip=mode, transpose=True)
+            MtU = tl.dot(unfold(tensor_cross, mode), tl.transpose(unfold(nn_core, mode)))
+            UtM = tl.transpose(MtU)
```

Comment:
If we use MtU, UtM, UtU, it would be nice to have a short Notes section in the docstring to briefly introduce the notation. We can also add a short mathematical bit in the user guide.

Reply:
I added mathematical notes to non_negative_tucker_hals function. I think these variables are mote clear now.